Eip micropayments standard for nfts and multi tokens (#4393)
ethereum/EIPs: EIPS/eip-micropayments_standard_for_nfts_and_multi_tokens.md

---

> eip: <to be assigned>

Reviewer: ```suggestion
eip: 4393
```

title: Micropayments Standard for NFTs and Multi Tokens
description: A standard interface for tip tokens that allows tipping to holders of NFTs and multi tokens

> author: Jules Lai (@julesl23), aka Julian Bushell

Reviewer: ```suggestion
author: Jules Lai (@julesl23)
```

Reviewer 2: Alternatively you can do `Julian Bushell (@jules123)`.  It is the author's choice what name they wish to go by.

discussions-to: https://ethereum-magicians.org/t/eip-proposal-micropayments-standard-for-nfts-and-multi-tokens/7366
status: Draft
type: Standards Track
category: ERC
created: 2021-10-24

> requires: 721, 1155, 165

Reviewer: ```suggestion
requires: 165, 721, 1155
```

---

## Abstract

> For the purpose of this proposal, a micropayment is termed as a financial transaction that involves usually a small sum of money, called "tips", that are sent for specific [ERC-721](https://eips.ethereum.org/EIPS/eip-721) NFTs and [ERC-1155](https://eips.ethereum.org/EIPS/eip-1155) multi tokens as rewards to their holders. A holder (also referred to as controller) is used here as a more generic term for owner in a semantic sense, as NFTs may represent non-digital assets, such as services etc.

Reviewer: Please use relative links of the form `[EIP-X](./eip-X.md)`

> This standard proposal outlines a generalised way to allow tipping via implementation of an ITipToken interface. The interface is intentionally kept to a minimum in order to allow for maximum use cases.

Reviewer: ```suggestion
This standard proposal outlines a generalised way to allow tipping via implementation of an `ITipToken` interface. The interface is intentionally kept to a minimum in order to allow for maximum use cases.
```

A user first deposits a compatible [ERC-20](https://eips.ethereum.org/EIPS/eip-20) to the tip token contract that is then held in escrow, in exchange for tip tokens. These tip tokens can then be sent by the user to NFTs and multi-tokens (that have been approved by the tip token contract for tipping) to be redeemed for the original ERC20 deposits on withdrawal by the holders as rewards.

Some use cases include:

- In game purchases and other virtual goods

- Tipping messages, posts, music and video content

- Donations/crowdfunding

- Distribution of royalties

- Pay per click advertising

- Incentivising use of services

- Reward cards and coupons

These can all leverage the security, immediacy and transparency of blockchain.

## Motivation

> A cheap way to send tips to any type of NFT or multi token. This can be achieved by gas optimising the tip token contract and sending the tips in batches using the tipBatch function from the interface.

Reviewer: ```suggestion
A cheap way to send tips to any type of NFT or multi token. This can be achieved by gas optimising the tip token contract and sending the tips in batches using the `tipBatch` function from the interface.
```
Please wrap code text in backticks.

    /// @param amount Amount of ERC20 token to withdraw as a reward
    function withdraw(uint256 amount) external payable;

    /// @notice MUST have idential behaviour to ERC20 balanceOf and is the amount
    /// of tip tokens held by 'user'
    /// @param user The user account
    /// @return The balance of tip tokens held by user
    function balanceOf(address user) external view returns (uint256);

    /// @notice The balance of deposit available to become rewards when
    /// user sends the tips
    /// @param user The user account
    /// @return The remaining balance of the ERC20 compatible token deposited
    function balanceDepositOf(address user) external view returns (uint256);

    /// @notice The amount of reward token owed to 'holder'
    /// @dev The pending tokens can come from the tip token contract account
    /// or from an external eschrow, depending on tip token implementation
    /// @param holder The holder of an NFT (NFT controller)
    /// @return The amount of reward tokens owed to the holder from tipping
    function rewardPendingOf(address holder) external view returns (uint256);
}
```

### Tip Token transfer and value calculations

> Tip token values are exchanged with ERC20 deposits and vice-versa. It is left to the tip token implementer to decide on the price of a tip token and hence how much tip to mint in exchange for the ERC20 deposited.

Reviewer: ```suggestion
Tip token values are exchanged with ERC-20 deposits and vice-versa. It is left to the tip token implementer to decide on the price of a tip token and hence how much tip to mint in exchange for the ERC-20 deposited.
```

Reviewer: Please update all occurrences of unhyphenated ERC/EIPs.

## Backwards Compatibility

A tip token contract can be fully compatible with ERC20 specification and inherit some functions such as transfer if the tokens are allowed to be sent directly to other users. Note that balanceOf has been adopted and MUST be the number of tips held by a user's address. If inheriting from, for example, OpenZeppelin's implementation of ERC20 token then their contract is responsible for maintaining the balance of tip token. Therefore, tip token balanceOf function SHOULD simply directly call the parent (super) contract's balanceOf function.

What hasn't been carried over to tip token standard, is the ability for a spender of other users' tips. For the moment, this standard does not foresee a need for this.

This EIP does not stress a need for tip token secondary markets or other use cases where identifying the tip token type with names rather than addresses might be useful, so these functions were left out of the ITipToken interface and is the remit for implementers.

> ## Reference Implementation
> 
> A [reference/example implementation and tests](https://github.com/julesl23/eip_msnmt) in the GitHub repository that can use be used to assist in implementing this specification.
> 

Reviewer 2: ```suggestion
```
External links should not be included in EIPs.  Either include an inline reference implementation, or files in `../assets/eip-4393/x.sol`.  Alternatively, this is an optional section and can simply be left out if that makes the most sense.

Author: > External links should not be included in EIPs. Either include an inline reference implementation, or files in `../assets/eip-4393/x.sol`. Alternatively, this is an optional section and can simply be left out if that makes the most sense.

All sorted in the latest commit. Let me know if there is anything else I need to do.

---
eip: 4393
title: Micropayments Standard for NFTs and Multi Tokens
description: A standard interface for tip tokens that allows tipping to holders of NFTs and multi tokens
author: Jules Lai (@julesl23)
discussions-to: https://ethereum-magicians.org/t/eip-proposal-micropayments-standard-for-nfts-and-multi-tokens/7366
status: Draft
type: Standards Track
category: ERC
created: 2021-10-24
requires: 165, 721, 1155
---

> ## Abstract

Reviewer: The abstract should be a short, technical description of the EIP. Please update this before moving forward into review.

Author: Thanks. Updated.

Some use cases include:

- In game purchases and other virtual goods

- Tipping messages, posts, music and video content

- Donations/crowdfunding

- Distribution of royalties

- Pay per click advertising

- Incentivising use of services

- Reward cards and coupons

These can all leverage the security, immediacy and transparency of blockchain.

## Motivation

A cheap way to send tips to any type of NFT or multi token. This can be achieved by gas optimising the tip token contract and sending the tips in batches using the `tipBatch` function from the interface.

To make it easy to implement into dapps a tipping service to reward the NFT and multi token holders. Allows for fairer distribution of revenue back to NFT holders from the user community.

To make the interface as minimal as possible in order to allow adoption into many different use cases.

## Specification

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

Reviewer: ```suggestion
```
Implied via EIP-1

Author: Done

> Smart contracts implementing the ERC-xxxx standard MUST implement all of the functions in the ERC-xxxx interface. MUST also emit the events specified in the interface so that a complete state of the tip token contract can be derived from the events emitted alone.

Reviewer: Need to replace EIP-xxxx -- usually prefer to not refer to the EIP itself by its own number. Instead say "this EIP" or a something similar.

Author: Done
